modules/lang-painless/src/main/java/org/elasticsearch/painless/lookup/PainlessLookupBuilder.java

```diff
@@ -122,157 +120,115 @@ public int hashCode() {
         }
 
-        String javaClassName = javaClass.getCanonicalName();
-        String painlessClassName = anyTypeNameToPainlessTypeName(javaClassName);
-        String importedPainlessClassName = anyTypeNameToPainlessTypeName(javaClassName.substring(javaClassName.lastIndexOf('.') + 1));
+        String canonicalClassName = clazz.getCanonicalName();
 
-        addPainlessClass(painlessClassName, importedPainlessClassName, javaClass, importPainlessClassName);
-    }
+        if (clazz.isArray()) {
+            throw new IllegalArgumentException("cannot add array type [" + canonicalClassName + "] as a class");
+        }
 
-    private void addPainlessClass(
-        String painlessClassName, String importedPainlessClassName, Class<?> javaClass, boolean importPainlessClassName) {
-        PainlessClassBuilder existingPainlessClassBuilder = javaClassesToPainlessClassBuilders.get(javaClass);
+        if (CLASS_NAME_PATTERN.matcher(canonicalClassName).matches() == false) {
+            throw new IllegalArgumentException("invalid class name [" + canonicalClassName + "]");
+        }
+
+        PainlessClassBuilder existingPainlessClassBuilder = classesToPainlessClasses.get(clazz);
 
         if (existingPainlessClassBuilder == null) {
-            PainlessClassBuilder painlessClassBuilder = new PainlessClassBuilder(painlessClassName, javaClass, Type.getType(javaClass));
-            painlessClassNamesToJavaClasses.put(painlessClassName, javaClass);
-            javaClassesToPainlessClassBuilders.put(javaClass, painlessClassBuilder);
-        } else if (existingPainlessClassBuilder.clazz.equals(javaClass) == false) {
-            throw new IllegalArgumentException("painless class [" + painlessClassName + "] illegally represents multiple java classes " +
-                    "[" + javaClass.getCanonicalName() + "] and [" + existingPainlessClassBuilder.clazz.getCanonicalName() + "]");
+            PainlessClassBuilder painlessClassBuilder = new PainlessClassBuilder(canonicalClassName, clazz, Type.getType(clazz));
+
+            canonicalTypeNamesToClasses.put(canonicalClassName, clazz);
+            classesToPainlessClasses.put(clazz, painlessClassBuilder);
+        } else if (existingPainlessClassBuilder.clazz.equals(clazz) == false) {
+            throw new IllegalArgumentException("class [" + canonicalClassName + "] " +
+                    "cannot represent multiple java classes with the same name from different class loaders");
         }
 
-        if (painlessClassName.equals(importedPainlessClassName)) {
-            if (importPainlessClassName == true) {
-                throw new IllegalArgumentException(
-                        "must use only_fqn parameter on painless class [" + painlessClassName + "] with no package");
+        String javaClassName = clazz.getName();
+        String importedCanonicalClassName = javaClassName.substring(javaClassName.lastIndexOf('.') + 1).replace('$', '.');
```

Comment:
I'm not sure it makes sense to remove the canonical.  It is a canonical name because it lops off the package but still follows the syntax with '.' for inner classes.  ```package.Class.InnerClass -> Class.InnerClass```.  A non-canonical name would be ```package.Class$InnerClass -> Class$InnerClass``` based on the formal java names.  What do you think?

Reply:
Ok sounds fine then.